### PR TITLE
599 remove manual qmlregistersingleton from the cpp source

### PR DESCRIPTION
**Pull request by DanielMcInnes**

### CODE REVIEW

File: src/backendconnection.cpp [hunk 1]
```diff
@@ -25,7 +25,7 @@
 namespace Victron {
 namespace VenusOS {
 
-BackendConnection* BackendConnection::instance(QQmlEngine *, QJSEngine *)
+BackendConnection* BackendConnection::create(QQmlEngine *, QJSEngine *)
```
Comment (DanielMcInnes) [inline src/backendconnection.cpp:28]: quirk #2: the factory function has to be named 'create', or it doesn't work.

File: src/language.h [hunk 2]
```diff
@@ -61,9 +62,11 @@ class LanguageModel : public QAbstractListModel
 class Language : public QObject
 {
 	Q_OBJECT
+	QML_ELEMENT
+	QML_SINGLETON
 	Q_PROPERTY(QLocale::Language current READ getCurrentLanguage WRITE setCurrentLanguage NOTIFY currentLanguageChanged FINAL)
 public:
-	explicit Language(QQmlEngine* engine);
+	explicit Language(QQmlEngine* engine = nullptr);
```
Comment (DanielMcInnes) [inline src/language.h:69]: quirk #3: you have to either use a static factory function named 'create', or do as we do here, and provide a default constructor, which wither takes no arguments, or provides default arguments. If we remove the '= nullptr' default argument, the singleton won't be accessible from qml.

File: src/main.cpp [hunk 3]
```diff
@@ -326,11 +326,7 @@ int main(int argc, char *argv[])
 	QObject::connect(&engine, &QQmlEngine::quit, &app, &QGuiApplication::quit);
 
 	/* Force construction of translator */
-	/*
-	int languageSingletonId = qmlTypeId("Victron.VenusOS", 2, 0, "Language");
-	Q_ASSERT(languageSingletonId);
-	(void)engine.singletonInstance<Victron::VenusOS::Language*>(languageSingletonId);
-	*/
+	Victron::VenusOS::Language language;
```
Comment (chriadam) [inline src/main.cpp:261]: You can't just construct a random one, you have to construct the engine's singleton:
```
Victron::VenusOS::Language* instance = engine.singletonInstance<Victron::VenusOS::Language*>(Victron::VenusOS::Language::typeId);
```

File: src/main.cpp [hunk 4]
```diff
@@ -244,12 +244,12 @@ void registerQmlTypes()
 		&Victron::VenusOS::Theme::instance);
 	qmlRegisterSingletonType<Victron::VenusOS::BackendConnection>(
 		"Victron.VenusOS", 2, 0, "BackendConnection",
-		&Victron::VenusOS::BackendConnection::create);*/
+		&Victron::VenusOS::BackendConnection::create);
 	qmlRegisterSingletonType<Victron::VenusOS::Language>(
 		"Victron.VenusOS", 2, 0, "Language",
 		[](QQmlEngine *engine, QJSEngine *) -> QObject* {
 			return new Victron::VenusOS::Language(engine);
```
Comment (chriadam) [inline src/main.cpp]: better to remove them altogether than simply comment them out, I think

File: ApplicationContent.qml [hunk 5]
```diff
@@ -40,9 +40,9 @@ Item {
 			if (pageManager.idleModeTimer.running) {
 				pageManager.idleModeTimer.restart()
 			}
-			if (pageManager.interactivity === VenusOS.PageManager_InteractionMode_Idle) {
+			if (pageManager.interactivity === Enums.PageManager_InteractionMode_Idle) {
```
Comment (chriadam) [inline ApplicationContent.qml:43]: this is fine, I think.  please remove the "ugh" from the commit message ;-)

File: src/main.cpp [hunk 6]
```diff
@@ -252,13 +252,13 @@ void registerQmlTypes()
 		});
 	qmlRegisterSingletonType<Victron::VenusOS::Enums>(
 		"Victron.VenusOS", 2, 0, "VenusOS",
-		&Victron::VenusOS::Enums::instance);*/
+		&Victron::VenusOS::Enums::instance);
 
 	qmlRegisterSingletonType<Victron::VenusOS::ActiveNotificationsModel>(
 		"Victron.VenusOS", 2, 0, "ActiveNotificationsModel",
 		[](QQmlEngine *, QJSEngine *) -> QObject * {
 		return Victron::VenusOS::ActiveNotificationsModel::instance();
```
Comment (chriadam) [inline src/main.cpp]: better to remove than comment out

File: CMakeLists.txt [hunk 7]
```diff
@@ -480,28 +480,46 @@ set (VENUS_QML_MODULE_SOURCES
...
+    src/clocktime.h
+    src/clocktime.cpp
+    src/uidhelper.h
+    src/uidhelper.cpp
+    src/frameratemodel.h
+    src/frameratemodel.cpp
+)
+
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "Emscripten")
+#    list(APPEND VenusQMLModule_CPP_SOURCES ${CONNMAN_SOURCES_MOCK} )
```
Comment (chriadam) [inline CMakeLists.txt]: intentionally commented out?

File: pages/SettingsPage.qml [hunk 8]
```diff
@@ -10,7 +10,7 @@
 
 import QtQuick
 import Victron.VenusOS
-import net.connman 0.1
+import net.connman 1.0
```
Comment (chriadam) [inline pages/SettingsPage.qml:13]: shouldn't need any version number

File: CMakeLists.txt [hunk 9]
```diff
@@ -684,40 +702,49 @@ qt_add_qml_module(Mqtt
 )
 # end Mqtt_QML_MODULE
 
-set(CONNMAN_SOURCES_MOCK
-    src/connman-api.h
-)
+# Connman module
 if("${CMAKE_SYSTEM_NAME}" STREQUAL "Emscripten")
-    list(APPEND SOURCES ${CONNMAN_SOURCES_MOCK} )
+    list(APPEND CONNMAN_SOURCES
```
Comment (chriadam) [inline CMakeLists.txt]: should this change be part of the next commit (the connman one)?

File: components/AcInputsCurrentLimits.qml [hunk 10]
```diff
@@ -42,23 +42,23 @@ Column {
 
 		delegate: ListButton {
 			text: Global.acInputs.currentLimitTypeToText(modelData.inputType)
-			writeAccessLevel: VenusOS.User_AccessType_User
+			writeAccessLevel: Enums.User_AccessType_User
```
Comment (chriadam) [inline components/AcInputsCurrentLimits.qml:45]: should these changes be part of the previous "Enums" commit?

File: CMakeLists.txt [hunk 11]
```diff
@@ -69,7 +69,7 @@ if(QT_KNOWN_POLICY_QTP0001)
     qt_policy(SET QTP0001 NEW) # >Qt6.5 only. Enabling this policy ensures that your QML module is placed under a default import path, and its types can be found without manual calls to QQmlEngine::addImportPath.
 endif()
 
-include_directories(. src src/veutil/inc src/veutil/inc/veutil/qt)
+include_directories(src/veutil/inc src .)
```
Comment (chriadam) [inline CMakeLists.txt:72]: I think this might be some churn, this was changed two commits ago I think?

File: src/main.cpp (hunk 3 above)
Comment (chriadam) [reply src/main.cpp:261]: See https://doc.qt.io/qt-6/qqmlengine.html#singletonInstance

File: CMakeLists.txt (hunk 9 above)
Comment (DanielMcInnes) [reply CMakeLists.txt]: probably, I'm going to squash all of the commits into 1 before I merge anyway.

File: pages/SettingsPage.qml (hunk 8 above)
Comment (DanielMcInnes) [reply pages/SettingsPage.qml:13]: that reminds me, I found a bug in qt_add_qml_module. If you have a version  0.1, you can't import it from qml. It has to be 1.0 or greater

File: src/main.cpp (hunk 3 above)
Comment (DanielMcInnes) [reply src/main.cpp]: ugh, I thought this magically worked, but it doesn't. Fixed

File: ApplicationContent.qml (hunk 5 above)
Comment (DanielMcInnes) [reply ApplicationContent.qml:43]: "ugh" removed

File: pages/SettingsPage.qml (hunk 8 above)
Comment (DanielMcInnes) [reply pages/SettingsPage.qml]: removed

File: src/main.cpp (hunk 4 above)
Comment (DanielMcInnes) [reply src/main.cpp]: done

File: src/main.cpp (hunk 6 above)
Comment (DanielMcInnes) [reply src/main.cpp]: done

File: CMakeLists.txt (hunk 7 above)
Comment (DanielMcInnes) [reply CMakeLists.txt]: yes, tidied up

File: src/main.cpp [hunk 12]
```diff
@@ -319,16 +313,18 @@ int main(int argc, char *argv[])
...
 	bool enableFpsCounter = false;
-	initBackend(&enableFpsCounter);
 
 	QQmlEngine engine;
+	initBackend(&enableFpsCounter);
 	engine.setProperty("colorScheme", Victron::VenusOS::Theme::Dark);
 	QObject::connect(&engine, &QQmlEngine::quit, &app, &QGuiApplication::quit);
 
 	/* Force construction of translator */
+	/*
```
Comment (chriadam) [inline src/main.cpp]: intentional in this commit?

File: CMakeLists.txt (hunk 7 above)
Comment (chriadam) [reply CMakeLists.txt]: in this commit?

File: CMakeLists.txt (hunk 9 above)
Comment (chriadam) [reply CMakeLists.txt]: I would prefer if they were kept separate, if possible

Comment (DanielMcInnes) [reply CMakeLists.txt]: this is fiddly to fix it up in history, do you want it done?

Comment (chriadam) [reply CMakeLists.txt]: yes please

File: src/main.cpp (hunk 4 above)
Comment (DanielMcInnes) [reply src/main.cpp]: done

File: CMakeLists.txt (hunk 7 above)
Comment (DanielMcInnes) [reply CMakeLists.txt]: I think this is good now

File: CMakeLists.txt (hunk 9 above)
Comment (DanielMcInnes) [reply CMakeLists.txt]: done

File: components/AcInputsCurrentLimits.qml (hunk 10 above)
Comment (DanielMcInnes) [reply components/AcInputsCurrentLimits.qml:45]: yes, done

File: CMakeLists.txt (hunk 11 above)
Comment (DanielMcInnes) [reply CMakeLists.txt]: fixed

File: src/main.cpp (hunk 12 above)
Comment (DanielMcInnes) [reply src/main.cpp]: done

**Review by chriadam — APPROVED** [review]
LGTM

**Pull request merged**